src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,700 @@
+            SendRouterSolicit();
+        }
+        ExitNow();
+    }
+
+    expireTime = GetPrefixExpireTime(aPio.GetValidLifetime());
+
+    if (!IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix) ||
+        (mDiscoveredOnLinkPrefixInvalidTimer.IsRunning() &&
+         expireTime > mDiscoveredOnLinkPrefixInvalidTimer.GetFireTime()))
+    {
+        mDiscoveredOnLinkPrefix.Set(aPio.GetPrefix(), aPio.GetPrefixLength());
+
+        otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
+                    mDiscoveredOnLinkPrefix.ToString().AsCString(), aPio.GetValidLifetime());
+
+        if (aPio.GetValidLifetime() == RouterAdv::kInfiniteLifetime)
+        {
+            mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+        }
+        else
+        {
+            mDiscoveredOnLinkPrefixInvalidTimer.FireAt(expireTime);
+        }
+
+        mRouterSolicitTimer.Stop();
+        needReevaluate = true;
+    }
+
+exit:
+    return needReevaluate;
+}
+
+TimeMilli RoutingManager::GetPrefixExpireTime(uint32_t aValidLifetime)
+{
+    if (aValidLifetime * 1000ULL + TimerMilli::GetNow().GetValue() < TimeMilli::kMaxDuration)
+    {
+        return Time(aValidLifetime * 1000 + TimerMilli::GetNow().GetValue());
+    }
+    else
+    {
+        return Time(Time::kMaxDuration);
+    }
+}
+
+void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
+                                               const uint8_t *     aBuffer,
+                                               uint16_t            aBufferLength)
+{
+    using RouterAdv::Option;
+    using RouterAdv::PrefixInfoOption;
+    using RouterAdv::RouterAdvMessage;
+
+    bool           needReevaluate = false;
+    const uint8_t *optionsBegin;
+    uint16_t       optionsLength;
+    const Option * option;
+
+    VerifyOrExit(aBufferLength >= sizeof(RouterAdvMessage));
+
+    OT_UNUSED_VARIABLE(aSrcAddress);
+    otLogInfoBr("received Router Advertisement from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                mInfraIfName);
+
+    optionsBegin  = aBuffer + sizeof(RouterAdvMessage);
+    optionsLength = aBufferLength - sizeof(RouterAdvMessage);
+
+    option = Option::GetNextOption(nullptr, optionsBegin, optionsLength);
+    while (option != nullptr)
+    {
+        if (option->GetType() == Option::Type::kPrefixInfo)
+        {
+            auto &      pio = *static_cast<const PrefixInfoOption *>(option);
+            Ip6::Prefix prefix;
+
+            prefix.Set(pio.GetPrefix(), pio.GetPrefixLength());
+            if (IsValidOnLinkPrefix(prefix))
+            {
+                needReevaluate |= HandlePrefixInfoOption(pio);
+            }
+            else
+            {
+                otLogInfoBr("ignore invalid prefix in PIO: %s", prefix.ToString().AsCString());
+            }
+        }
+        option = Option::GetNextOption(option, optionsBegin, optionsLength);
+    }
+
+    if (needReevaluate)
+    {
+        EvaluateRoutingPolicy();
```

Comment:
We seems to call `EvaluateRoutingPolicy` when the first RA is received. But should we wait for a while (e.g. 1~3s) after RS to collect all RAs?

We might also need to use send RS for better confidence just like what we did in DUA DAD?

Reply:
> We seems to call EvaluateRoutingPolicy when the first RA is received. But should we wait for a while (e.g. 1~3s) after RS to collect all RAs?

What about unsolicited RAs? Would it be much simpler to just evaluate our routing policy on every single RA? In most cases, only the first RA will change our routing policy.

Reply:
Agree that we can start `EvaluateRoutingPolicy` once we found a on link prefix.